### PR TITLE
Add MSSQLServer Vertical and Horizontal Scaling Docs

### PR DESCRIPTION
**Pull request by Neaj-Morshad-101**

### CODE REVIEW

File: docs/guides/mssqlserver/scaling/horizontal-scaling/overview.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - [MSSQLServerOpsRequest](/docs/guides/mssqlserver/concepts/opsrequest.md)
+
+## How Horizontal Scaling Process Works
+
+The following diagram shows how `KubeDB` Ops Manager used to scale up the number of members of a `MSSQLServer` cluster. Open the image in a new tab to see the enlarged version.
+
+<figure align="center">
+  <img alt="Horizontal scaling Flow" src="/docs/images/day-2-operation/mssqlserver/ms-horizontal-scaling.svg">
+<figcaption align="center">Fig: Horizontal scaling process of MSSQLServer</figcaption>
+</figure>
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/horizontal-scaling/overview.md:32]: image's indication should be proper

File: docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+```
+
+**Wait for the cluster to be ready:**
+
+`KubeDB` operator watches for `MSSQLServer` objects using Kubernetes API. When a `MSSQLServer` object is created, `KubeDB` operator will create a new PetSet, Services, and Secrets, etc. A secret called `mssql-ag-cluster-auth` (format: <em>{mssqlserver-object-name}-auth</em>) will be created storing the password for mssqlserver superuser.
+Now, watch `MSSQLServer` is going to `Running` state and also watch `PetSet` and its pod is created and going to `Running` state,
+
+```bash
+$ watch kubectl get ms,petset,pods -n demo
+Every 2.0s: kubectl get ms,petset,pods -n demo
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md:164]: remove this line

File: docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+
+**Verify Scale-Up Succeeded:**
+
+If everything goes well, `KubeDB` Ops Manager will scale up the PetSet's `Pod`. After the scaling process is completed successfully, the `KubeDB` Ops Manager updates the replicas of the `MSSQLServer` object.
+
+First, we will wait for `MSSQLServerOpsRequest` to be successful. Run the following command to watch `MSSQLServerOpsRequest` cr,
+
+```bash
+$ watch kubectl get mssqlserveropsrequest -n demo msops-hscale-up
+Every 2.0s: kubectl get mssqlserveropsrequest -n demo msops-hscale-up
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md:275]: remove

File: docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md [hunk 4]
```diff
@@ -0,0 +1,717 @@
...
+
+**Verify Scale-down Succeeded:**
+
+If everything goes well, `KubeDB` Ops Manager will scale down the PetSet's `Pod`. After the scaling process is completed successfully, the `KubeDB` Ops Manager updates the replicas of the `MSSQLServer` object.
+
+Now, we will wait for `MSSQLServerOpsRequest` to be successful. Run the following command to watch `MSSQLServerOpsRequest` cr,
+
+```bash
+$ watch kubectl get mssqlserveropsrequest -n demo msops-hscale-down
+Every 2.0s: kubectl get mssqlserveropsrequest -n demo msops-hscale-down
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md:519]: remove

File: docs/guides/mssqlserver/scaling/vertical-scaling/overview.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  - [MSSQLServerOpsRequest](/docs/guides/mssqlserver/concepts/opsrequest.md)
+
+## How Vertical Scaling Process Works
+
+The following diagram shows how the `KubeDB` Ops Manager used to update the resources of the `MSSQLServer`. Open the image in a new tab to see the enlarged version.
+
+<figure align="center">
+  <img alt="Vertical scaling process of MSSQLServer" src="/docs/images/day-2-operation/mssqlserver/ms-vertical-scaling.svg">
+<figcaption align="center">Fig: Vertical scaling process of MSSQLServer</figcaption>
+</figure>
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/vertical-scaling/overview.md:32]: image fix

File: docs/guides/mssqlserver/scaling/vertical-scaling/overview.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+6. When it finds one, it halts the `MSSQLServer` object so that the `KubeDB` provisioner operator doesn't perform any operation on the `MSSQLServer` during the scaling process.
+
+7. Then the KubeDB Ops-manager operator will update resources of the PetSet's Pods to reach desired state.
+
+8. After successful updating of the resources of the PetSet's Pods, the `KubeDB` Ops Manager updates the `MSSQLServer` object resources to reflect the updated state.
+
+9. After successful updating of the `MSSQLServer` resources, the `KubeDB` Ops Manager resumes the `MSSQLServer` object so that the `KubeDB` Provisioner operator resumes its usual operations.
+
+In the next doc, we are going to show a step-by-step guide on updating resources of MSSQLServer database using vertical scaling operation.
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/vertical-scaling/overview.md:54]: add "Next Steps"

File: docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+`KubeDB` watches for `MSSQLServer` objects using Kubernetes API. When a `MSSQLServer` object is created, `KubeDB` will create a new PetSet, Services, and Secrets, etc.
+Now, watch `MSSQLServer` is going to be in `Running` state and also watch `PetSet` and its pod is created and going to be in `Running` state,
+
+
+
+
+
+```bash
+$ watch kubectl get ms,petset,pods -n demo
+Every 2.0s: kubectl get ms,petset,pods -n demo
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md:139]: remove

File: docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+
+**Verify MSSQLServer resources updated successfully:**
+
+If everything goes well, `KubeDB-Ops-Manager` will update the resources of the PetSet's `Pod` containers. After a successful scaling process is done, the `KubeDB-Ops-Manager` updates the resources of the `MSSQLServer` object.
+
+First, we will wait for `MSSQLServerOpsRequest` to be successful. Run the following command to watch `MSSQLServerOpsRequest` CR,
+
+```bash
+$ watch kubectl get mssqlserveropsrequest -n demo mops-vscale-standalone
+Every 2.0s: kubectl get mssqlserveropsrequest -n demo mops-vscale-standalone
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md:217]: remove

File: docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+To clean up the Kubernetes resources created by this tutorial, run:
+
+```bash
+kubectl delete mssqlserver -n demo mssql-standalone
+kubectl delete mssqlserveropsrequest -n demo mops-vscale-standalone
+kubectl delete issuer -n demo mssqlserver-ca-issuer
+kubectl delete secret -n demo mssqlserver-ca
+kubectl delete ns demo
+```
+
```
Comment (takiaesha) [inline docs/guides/mssqlserver/scaling/vertical-scaling/standalone.md:345]: add next steps if possible

**Review by takiaesha — CHANGES_REQUESTED** [review]

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+## What is MSSQLServerOpsRequest
+
+`MSSQLServerOpsRequest` is a Kubernetes `Custom Resource Definitions` (CRD). It provides declarative configuration for [Microsoft SQL Server](https://learn.microsoft.com/en-us/sql/sql-server/) administrative operations like database version updating, horizontal scaling, vertical scaling, reconfigure, volume expansion, etc. in a Kubernetes native way.
+
+## MSSQLServerOpsRequest CRD Specifications
+
+Like any official Kubernetes resource, a `MSSQLServerOpsRequest` has `TypeMeta`, `ObjectMeta`, `Spec` and `Status` sections.
+
+Here, some sample `MSSQLServerOpsRequest` CRs for different administrative operations is given below,
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md]: operations are given below

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+Here, we are going to describe the various sections of a `MSSQLServerOpsRequest` CR.
+
+### MSSQLServerOpsRequest `Spec`
+
+A `MSSQLServerOpsRequest` object has the following fields in the `spec` section.
+
+#### spec.databaseRef
+
+`spec.databaseRef` is a required field that point to the [MSSQLServer](/docs/guides/mssqlserver/concepts/mssqlserver.md) object where the administrative operations will be applied. This field consists of the following sub-field:
+
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md:120]: following sub-fields

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+`status.observedGeneration` shows the most recent generation observed by the `MSSQLServerOpsRequest` controller.
+
+#### status.conditions
+
+`status.conditions` is an array that specifies the conditions of different steps of `MSSQLServerOpsRequest` processing. Each condition entry has the following fields:
+
+- `types` specifies the type of the condition. MSSQLServerOpsRequest has the following types of conditions:
+
+| Type                | Meaning                                                                                   |
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md]: this table is not correctly alligned

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+`MSSQLServerOpsRequest` is a Kubernetes `Custom Resource Definitions` (CRD). It provides declarative configuration for [Microsoft SQL Server](https://learn.microsoft.com/en-us/sql/sql-server/) administrative operations like database version updating, horizontal scaling, vertical scaling, reconfigure, volume expansion, etc. in a Kubernetes native way.
+
+## MSSQLServerOpsRequest CRD Specifications
+
+Like any official Kubernetes resource, a `MSSQLServerOpsRequest` has `TypeMeta`, `ObjectMeta`, `Spec` and `Status` sections.
+
+Here, some sample `MSSQLServerOpsRequest` CRs for different administrative operations is given below,
+
+Sample `MSSQLServerOpsRequest` for updating database:
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md:27]: MSSQLServerOpsRequest CR
same changes for other places

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+If you want to update your MSSQLServer version, you have to specify the `spec.updateVersion`  section that specifies the desired version information. This field consists of the following sub-field:
+
+- `spec.updateVersion.targetVersion` refers to a [MSSQLServerVersion](/docs/guides/mssqlserver/concepts/catalog.md) CR that contains the MSSQLServer version information where you want to update.
+
+>You can only update between MSSQLServer versions. KubeDB does not support downgrade for MSSQLServer.
+
+#### spec.horizontalScaling
+
+If you want to scale-up or scale-down your MSSQLServer cluster, you have to specify `spec.horizontalScaling` section. This field consists of the following sub-field:
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md:145]: sub-fields

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+
+- `spec.verticalScaling.exporter` indicates the `exporter` container resources. It has the same structure as `spec.verticalScaling.mssqlserver` and you can scale the resource the same way as `MSSQLServer` container.
+
+>You can increase/decrease resources for both `MSSQLServer` container and `exporter` container on a single `MSSQLServerOpsRequest` CR.
+
+### spec.volumeExpansion
+
+> To use the volume expansion feature the storage class must support volume expansion
+
+If you want to expand the volume of your MSSQLServer cluster, you have to specify `spec.volumeExpansion` section. This field consists of the following sub-field:
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md]: sub-fields

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+|---------------------| ----------------------------------------------------------------------------------------- |
+| `Progressing`       | Specifies that the operation is now progressing                                           |
+| `Successful`        | Specifies such a state that the operation on the database has been successful.            |
+| `HaltDatabase`      | Specifies such a state that the database is halted by the operator                        |
+| `ResumeDatabase`    | Specifies such a state that the database is resumed by the operator                       |
+| `Failed`           | Specifies such a state that the operation on the database has been failed.                |
+| `Scaling`           | Specifies such a state that the scaling operation on the database has stared              |
+| `VerticalScaling`   | Specifies such a state that vertical scaling has performed successfully on database       |
+| `HorizontalScaling` | Specifies such a state that horizontal scaling has performed successfully on database     |
+| `Updating`          | Specifies such a state that database updating operation has stared                       |
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md]: started

File: docs/guides/mssqlserver/concepts/opsrequest.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+- `types` specifies the type of the condition. MSSQLServerOpsRequest has the following types of conditions:
+
+| Type                | Meaning                                                                                   |
+|---------------------| ----------------------------------------------------------------------------------------- |
+| `Progressing`       | Specifies that the operation is now progressing                                           |
+| `Successful`        | Specifies such a state that the operation on the database has been successful.            |
+| `HaltDatabase`      | Specifies such a state that the database is halted by the operator                        |
+| `ResumeDatabase`    | Specifies such a state that the database is resumed by the operator                       |
+| `Failed`           | Specifies such a state that the operation on the database has been failed.                |
+| `Scaling`           | Specifies such a state that the scaling operation on the database has stared              |
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/concepts/opsrequest.md]: started

File: docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md [hunk 18]
```diff
@@ -0,0 +1,717 @@
...
+  Warning  get pod; ConditionStatus:False                                              89s   KubeDB Ops-manager Operator  get pod; ConditionStatus:False
+  Warning  get pod; ConditionStatus:True; PodName:mssql-ag-cluster-2                   59s   KubeDB Ops-manager Operator  get pod; ConditionStatus:True; PodName:mssql-ag-cluster-2
+  Warning  delete pvc; ConditionStatus:True; PodName:mssql-ag-cluster-2                59s   KubeDB Ops-manager Operator  delete pvc; ConditionStatus:True; PodName:mssql-ag-cluster-2
+  Warning  get pvc; ConditionStatus:False                                              59s   KubeDB Ops-manager Operator  get pvc; ConditionStatus:False
+  Warning  get pod; ConditionStatus:True; PodName:mssql-ag-cluster-2                   54s   KubeDB Ops-manager Operator  get pod; ConditionStatus:True; PodName:mssql-ag-cluster-2
+  Warning  delete pvc; ConditionStatus:True; PodName:mssql-ag-cluster-2                54s   KubeDB Ops-manager Operator  delete pvc; ConditionStatus:True; PodName:mssql-ag-cluster-2
+  Warning  get pvc; ConditionStatus:True                                               54s   KubeDB Ops-manager Operator  get pvc; ConditionStatus:True
```
Comment (souravbiswassanto) [inline docs/guides/mssqlserver/scaling/horizontal-scaling/mssqlserver.md:655]: remove extra events

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

**Review by souravbiswassanto — APPROVED** [review]

**Pull request merged**